### PR TITLE
[Partially Complete] Solana token stream protocol

### PR DESCRIPTION
**Pull request by SushantChandla**
Thanks for submitting this tutorial for Figment Learn!

Please confirm the following before submitting your tutorial for review - 

- [X] My tutorial follows the [Tutorial Contribution Guidelines](https://www.notion.so/figmentnetworks/Figment-Learn-Contributor-Guidelines-b08691366e144d33a5636d9276a44ec8).

In particular:

- [ ] The title of the Pull Request matches the title of my tutorial.
- [X] I've followed the [tutorial template](https://www.notion.so/figmentnetworks/Sample-Tutorial-Structure-667ac2aad36f4f94a3ebaca053180b2d) structure.
- [ ] I've [proofread the tutorial](https://grammarly.com) for spelling and grammar mistakes.
- [X] I've run a linter & prettifier to format my code blocks.
- [ ] I've thoroughly explained the code, with inline code comments and text explanations.
- [ ] I've run the tutorial to test it and the code works.

**NOTE**: We cannot review your Pull Request if it does not meet the standard outlined above.

### CODE REVIEW

**Comment by pyr0gan** [comment]
Hi @SushantChandla,

What is your timeline for completing this tutorial? 😄

**Comment by pyr0gan** [comment]
Hi @SushantChandla,

You will need to update this Pull Request to work with the new directory structure of the repository before it can be merged:

- Move images to the `assets` directory, you can then change all image links to the relative path `../assets/image-filename.png`
- Move the markdown file into the `solana` directory, so the new path would be `solana/solana-streaming-protocol.md`

**Comment by SushantChandla** [comment]
Okay, I will do that.

**Comment by SushantChandla** [comment]
Hey @pyr0gan,
It's finally complete, Kindly review it.

**Comment by pyr0gan** [comment]
Hi @SushantChandla! My sincere apologies, I was wrong about the image links. It is best to use a direct URL format to avoid any broken images. If you could please change the image links to use `https://github.com/figment-networks/learn-tutorials/raw/master/assets/image-filename.png` instead of the relative path, we can be confident that all images will display properly in the tutorial 😅 .

File: solana/solana-streaming-protocol.md [hunk 1]
```diff
@@ -0,0 +1,1537 @@
...
+
+{% embed url="https://youtu.be/r9CwLIR4afY" caption="UI Sample"  %}
+
+Now let us write the functions to connect with the sol-stream-program. 
+
+In `src/actions/index.js`:
+
+```js
+import {
+	SystemProgram,
```
Comment (pyr0gan) [inline solana/solana-streaming-protocol.md:1226]: It would be better to use spaces instead of tabs here. Overall, if you are using prettier for code formatting, JSX should have 2 spaces per level of indentation.

File: solana/solana-streaming-protocol.md [hunk 2]
```diff
@@ -0,0 +1,1537 @@
...
+import axios from "axios";
+
+export default  axios.create({
+  baseURL: "http://127.0.0.1:8000",
+});
+```
+
+Now that our `baseURL` is set we can write the function in `src/actions/index.js` update the `getAllStreams` function to:
+
+```js
```
Comment (pyr0gan) [inline solana/solana-streaming-protocol.md:1262]: Reduce the indentation in these code blocks by using 2 spaces per level of indentation (run prettier on the code before pasting it into the markdown file). Right now there is too much indentation:
Screen Shot 2021-11-18 at 4 01 55 PM

File: solana/solana-streaming-protocol.md [hunk 3]
```diff
@@ -0,0 +1,1537 @@
...
+			dispatch({
+				type: "DATA_NOT_RECEIVED",
+				result: { data: null },
+			})
+		}
+	};
+};
+```
+
+In this we requesting "/\<pubkey>" and then we dispatch the result. We have added the code in the try and catch block to handle network errors.
```
Comment (pyr0gan) [inline solana/solana-streaming-protocol.md:1287]: > We are requesting the route "/\<pubkey>" and then we dispatch the result. We have added the code in the try and catch block to handle network errors.

File: solana/solana-streaming-protocol.md [hunk 4]
```diff
@@ -0,0 +1,1537 @@
...
+
+## Create Stream
+Now let's write the functions that will contain the instructions to our program. We will also create a Schema for data Serialization.
+
+```js
```
Comment (pyr0gan) [inline solana/solana-streaming-protocol.md:1292]: Fix the indentation of all code blocks  😄

File: solana/solana-streaming-protocol.md [hunk 5]
```diff
@@ -0,0 +1,1537 @@
...
+				['receiver', [32]],
+				['lamports_withdrawn', 'u64'],
+				['amount_second', 'u64']]
+		}]]);
+}
+```
+
+For creating stream instruction we have created a Schema that describes the data type for variables. We will be using this in the `createStream` function:
+
+```js
```
Comment (pyr0gan) [inline solana/solana-streaming-protocol.md:1314]: Fix indentation -> 2 spaces per level of indentation 👍

File: solana/solana-streaming-protocol.md [hunk 6]
```diff
@@ -0,0 +1,1537 @@
...
+				amount_second: amountSpeed,
+			});
+
+            let data = serialize(CreateStreamInput.schema, create_stream_input);
+```
+
+In the code above we have created a pubkey which we will use to create a Program Drived account(PDA). We have used the function `serialize` from the borsh package to get the Uint8Array of the input data.
+
+
+```js
```
Comment (pyr0gan) [inline solana/solana-streaming-protocol.md:1344]: Definitely too much indentation here 😅

File: solana/solana-streaming-protocol.md [hunk 7]
```diff
@@ -0,0 +1,1537 @@
...
+			let data_to_send = new Uint8Array([1, ...data]);
+
+            let rent = await connection.getMinimumBalanceForRentExemption(96);
+```
+
+We have to append `1` in our Uint8Array as we are using the first character as the tag. Please refer to the unpack function in `instructions.rs`.
+
+And then we can get the minimum lamports required to make the account lamports by using the provided function on the connection variable.
```
Comment (pyr0gan) [inline solana/solana-streaming-protocol.md:1352]: > We can then get the minimum lamports required to make the account by using the `getMinimumBalanceForRentExemption` function on the connection variable.

File: solana/solana-streaming-protocol.md [hunk 8]
```diff
@@ -0,0 +1,1537 @@
...
+> Note: I checked the size of the data ( the value 96) by a test [HERE](https://github.com/SushantChandla/sol-stream-program/blob/d0d537889b155d16fa3c0879e61e4f68c3d47f07/src/state.rs#L42).
+
+```js
```
Comment (pyr0gan) [inline solana/solana-streaming-protocol.md:1355]: Fix indentation 😄

File: solana/solana-streaming-protocol.md (hunk 4 above)
Comment (pyr0gan) [reply solana/solana-streaming-protocol.md:1292]: Which file does this code belong in?

File: solana/solana-streaming-protocol.md (hunk 8 above)
Comment (pyr0gan) [reply solana/solana-streaming-protocol.md:1355]: I'm a little lost.. Is this still in `src/actions/index.js`?

File: solana/solana-streaming-protocol.md [hunk 9]
```diff
@@ -0,0 +1,1537 @@
...
+					{ pubkey: wallet.publicKey, isSigner: true, },
+					{ pubkey: receiverAddress, isSigner: false, },
+					{ pubkey: adminAddress, isSigner: false, isWritable: true }
+				],
+				programId: programAccount,
+				data: data_to_send
+			});
+```
+
+Here we have created 2 instruction
```
Comment (pyr0gan) [inline solana/solana-streaming-protocol.md:1378]: > Here we have created 2 instructions :

note: Where is "here"? It helps to indicate which file is being edited.

File: solana/solana-streaming-protocol.md [hunk 10]
```diff
@@ -0,0 +1,1537 @@
...
+
+1. To the system program to create a Program drived account with the space of `96` and lamports which are equal to `0.03(admin cut)+solana rent+total amount user want to send`.
```
Comment (pyr0gan) [inline solana/solana-streaming-protocol.md:1380]: > 1. To the system program to create a Program Derived Account with the space of `96` and lamports which are equal to (the admin cut) + 0.03 SOL (the amount the user wants to send) + solana rent

note: This could use some clarification.

File: solana/solana-streaming-protocol.md [hunk 11]
```diff
@@ -0,0 +1,1537 @@
...
+2. We have created an instruction to our Program we have passed the account that will be used which corresponds to `accounts: &[AccountInfo],` in our program in `keys`, `programId` is the address we got when we deployed our program and `data` is equal to `data_to_send` variable.
```
Comment (pyr0gan) [inline solana/solana-streaming-protocol.md:1381]: This is difficult to understand. I know what you're trying to say, but it would not be obvious to most readers. It would be better expressed as

> This instruction contains an array of public keys associated with the transaction, specifically the `newAccount` (the PDA being created), the signer's public key coming from their connected wallet, and also the receiver and admin addresses. The `programId` is the deployed program address and `data` is the instruction data which includes the tag to specify which instruction to execute (CreateStream, WithdrawFromStream or CloseStream).

File: solana/solana-streaming-protocol.md [hunk 12]
```diff
@@ -0,0 +1,1537 @@
...
+
+Then we create a transaction object we have used the `setPayerAndBlockhashTransaction` function which we will add later.
+Once we have the transaction object we have to send the transaction with the help of the wallet. We have then dispatched the result.
+In case of an error, we can just alert the user.
+
+```js
```
Comment (pyr0gan) [inline solana/solana-streaming-protocol.md:1387]: Fix the indentation 😄

File: solana/solana-streaming-protocol.md [hunk 13]
```diff
@@ -0,0 +1,1537 @@
...
+	return transaction;
+}
+```
+
+## Withdraw
+Now let's write the `withdraw` function for this one also we have created the `WithdrawInput` class with schema:
+
+We only have the `amount` variable in the struct for withdrawal input.
+
+```js
```
Comment (pyr0gan) [inline solana/solana-streaming-protocol.md:1432]: Fix indentation 👍

File: solana/solana-streaming-protocol.md [hunk 14]
```diff
@@ -0,0 +1,1537 @@
...
+				['amount', 'u64'],
+			]
+		}]]);
+}
+```
+
+This time we aren't creating a PDA so we only have one instruction in this transaction which is to our program.
+Note that the `streamId` is the address to the `PDA` account.
+
+Here in `data_to_send`, we have to append 2 as you can see in our program's `src/instrction.rs` file we have `tag` 2 for withdraw.
```
Comment (pyr0gan) [inline solana/solana-streaming-protocol.md:1452]: > Here in `data_to_send`, we have to specify 2 at the beginning of the array. As you recall, in our program's `instruction.rs` file we have set the tag 2 for the WithdrawFromStream instruction.

File: solana/solana-streaming-protocol.md [hunk 15]
```diff
@@ -0,0 +1,1537 @@
...
+
+```js
```
Comment (pyr0gan) [inline solana/solana-streaming-protocol.md:1454]: Fix indentation 👍

File: solana/solana-streaming-protocol.md [hunk 16]
```diff
@@ -0,0 +1,1537 @@
...
+```
+
+We have dispatched the withdraw response in the end.
+
+## Close Stream
+Now let's write the last function in the file to close a stream.
+
+For this function, we do not have any instruction data we are only passing the tag. Then we have packed the instruction by calling `setPayerAndBlockhashTransaction` and then dispatched the result.
+
+```js
```
Comment (pyr0gan) [inline solana/solana-streaming-protocol.md:1497]: Fix indentation 👍

File: solana/solana-streaming-protocol.md [hunk 17]
```diff
@@ -0,0 +1,1537 @@
...
+				['start_time', 'u64'],
+				['end_time', 'u64'],
+				['receiver', [32]],
+				['lamports_withdrawn', 'u64'],
+				['amount_second', 'u64']]
+		}]]);
+}
+```
+
+For creating stream instruction we have created a Schema that describes the data type for variables. We will be using this in the `createStream` function:
```
Comment (pyr0gan) [inline solana/solana-streaming-protocol.md:1312]: > For creating the stream instructions we have defined a Schema that describes the data type for variables. We will be using this in the `createStream` function:

**Review by pyr0gan — CHANGES_REQUESTED** [review]
Hi @SushantChandla!

Thanks for writing this awesome tutorial, I bet the community will have a lot of fun with it! There are some fixes to be made regarding grammar, and code block indentation.

For Rust, 4 spaces per level of indentation is OK. Tabs are also acceptable _in general_, but to keep the code blocks readable without scrolling we prefer 4 spaces.

For JavaScript, `prettier` settings will provide a default Tab Width of 2. It's a bit of a semantic nightmare, as tabs are still used for indentation but spaces are used to align things 🤷. 
Long story short, we're trying to fit things in a fixed width column on the page. More indentation makes it difficult to read a code block without horizontal scrolling. I'm sure you see where I'm coming from 😄 

Have a look over my comments and make the necessary changes. If you have any questions or concerns, please tag me in your reply so I am notified 🙏

File: solana/solana-streaming-protocol.md [hunk 18]
```diff
@@ -0,0 +1,1537 @@
...
+This will create a migrations folder with the name `<Timestamp>_Stream` with `down.sql` and `up.sql` files.
+
+Update the `up.sql` file to:
+
+```sql
+-- Your SQL goes here
+
+Create Table streams(
+   pda_account Varchar PRIMARY KEY,
+    start_time BIGINT NOT NUll,
```
[the remark this answers is not in the document]
Comment (SushantChandla) [reply solana/solana-streaming-protocol.md:823]: The SQL Keywords are case-insensitive, but `NUll` sure looks wired. Updated!

**Comment by SushantChandla** [comment]
I added all the requested changes, Please review @pyr0gan.

**Review by pyr0gan — APPROVED** [review]
Hi @SushantChandla, 

Thank you for taking care of the changes, the tutorial is easier to follow now and flows nicely don't you think? 😄 🥳

**Comment by SushantChandla** [comment]
Yes it is. Thanks for all the suggestions @pyr0gan.  🥳
